File: apps/desktop/src-tauri/src/daemon_client.rs
//! Relays one IPC envelope from the UI to lawd over the authenticated
//! Unix-domain socket. The shell learns the socket path and bearer token from
//! the daemon's handshake file; it never holds provider, file, or credential
//! access of its own. Blocking socket I/O runs on a blocking task so the async
//! command does not stall the UI thread.

use std::env;
use std::fs;
use std::io::{BufRead, BufReader, Write};
use std::os::unix::net::UnixStream;
use std::path::PathBuf;
use std::time::Duration;

use serde_json::Value;

fn runtime_dir() -> PathBuf {
    let base = match env::var("XDG_RUNTIME_DIR") {
        Ok(v) if !v.trim().is_empty() => PathBuf::from(v),
        _ => env::temp_dir(),
    };
    base.join("law")
}

fn handshake_path() -> PathBuf {
    runtime_dir().join("lawd.json")
}

/// Error envelope matching the shared contract, so the UI validates it uniformly.
fn error_envelope(request: &Value, code: &str, message: &str, recovery: &str) -> Value {
    serde_json::json!({
        "protocol": 1,
        "id": request.get("id").and_then(Value::as_str).unwrap_or("unknown"),
        "op": request.get("op").and_then(Value::as_str).unwrap_or("unknown"),
        "schemaVersion": request.get("schemaVersion").and_then(Value::as_u64).unwrap_or(1),
        "ok": false,
        "error": { "code": code, "message": message, "recovery": recovery }
    })
}

/// Blocking relay: read handshake, connect, send `{token,request}\n`, read one
/// `{response}` line. Returns the daemon's response envelope, or a typed error
/// envelope when the daemon is unreachable.
pub fn relay(request: Value) -> Value {
    let handshake = match fs::read_to_string(handshake_path()) {
        Ok(s) => s,
        Err(_) => {
            return error_envelope(
                &request,
                "UNAVAILABLE",
                "Aster daemon is not running (no handshake file)",
                "start lawd and retry",
            )
        }
    };
    let hs: Value = match serde_json::from_str(&handshake) {
        Ok(v) => v,
        Err(_) => return error_envelope(&request, "INTERNAL", "invalid daemon handshake", "restart lawd"),
    };
    let socket_path = hs.get("socketPath").and_then(Value::as_str).unwrap_or_default();
    let token = hs.get("token").and_then(Value::as_str).unwrap_or_default();
    if socket_path.is_empty() || token.is_empty() {
        return error_envelope(&request, "INTERNAL", "daemon handshake missing socket or token", "restart lawd");
    }

    let stream = match UnixStream::connect(socket_path) {
        Ok(s) => s,
        Err(_) => {
            return error_envelope(
                &request,
                "UNAVAILABLE",
                "could not connect to the Aster daemon socket",
                "start lawd and retry",
            )
        }
    };
    let _ = stream.set_read_timeout(Some(Duration::from_secs(30)));
    let _ = stream.set_write_timeout(Some(Duration::from_secs(30)));

    let frame = serde_json::json!({ "token": token, "request": request });
    let mut writer = stream.try_clone().expect("clone unix stream");
    if writeln!(writer, "{}", frame).is_err() {
        return error_envelope(&request, "UNAVAILABLE", "failed to send request to daemon", "retry");
    }

    let mut reader = BufReader::new(stream);
    let mut line = String::new();
    if reader.read_line(&mut line).is_err() || line.trim().is_empty() {
        return error_envelope(&request, "UNAVAILABLE", "no response from daemon", "retry");
    }
    match serde_json::from_str::<Value>(&line) {
        Ok(v) => v.get("response").cloned().unwrap_or_else(|| {
            error_envelope(&request, "INTERNAL", "daemon returned an unframed response", "retry")
        }),
        Err(_) => error_envelope(&request, "INTERNAL", "daemon returned invalid JSON", "retry"),
    }
}

File: src/mcp/daemon-client.ts
import { readFileSync } from "node:fs";
import { createConnection } from "node:net";
import { tmpdir } from "node:os";
import { join } from "node:path";

interface Handshake { socketPath: string; token: string; protocol: number }
interface ResponseFrame { response?: { ok?: boolean; result?: unknown; error?: { message?: string; recovery?: string } } }

function handshakePath(): string {
  const base = process.env.XDG_RUNTIME_DIR?.trim() || tmpdir();
  return join(base, "law", "lawd.json");
}

function readHandshake(): Handshake {
  try {
    const value = JSON.parse(readFileSync(handshakePath(), "utf8")) as Handshake;
    if (!value.socketPath || !value.token || !value.protocol) throw new Error("invalid handshake");
    return value;
  } catch {
    throw new Error("Aster is not running. Open the Aster desktop application and try again.");
  }
}

export async function callAsterDaemon<T>(op: string, payload: unknown, timeoutMs = 15_000): Promise<T> {
  const handshake = readHandshake();
  const request = {
    protocol: handshake.protocol,
    id: `mcp-${Date.now()}-${crypto.randomUUID().slice(0, 8)}`,
    op,
    schemaVersion: 1,
    payload,
  };
  const frame = await new Promise<ResponseFrame>((resolve, reject) => {
    const socket = createConnection(handshake.socketPath);
    let buffer = "";
    const timer = setTimeout(() => { socket.destroy(); reject(new Error(`Aster IPC timed out during ${op}.`)); }, timeoutMs);
    const finish = (fn: () => void) => { clearTimeout(timer); socket.destroy(); fn(); };
    socket.setEncoding("utf8");
    socket.on("connect", () => socket.write(`${JSON.stringify({ token: handshake.token, request })}\n`));
    socket.on("data", (chunk: string) => {
      buffer += chunk;
      const newline = buffer.indexOf("\n");
      if (newline < 0) return;
      try { const parsed = JSON.parse(buffer.slice(0, newline)) as ResponseFrame; finish(() => resolve(parsed)); }
      catch { finish(() => reject(new Error("Aster returned invalid IPC data."))); }
    });
    socket.on("error", (error) => finish(() => reject(error)));
  });
  if (!frame.response?.ok) {
    const error = frame.response?.error;
    throw new Error([error?.message ?? `Aster operation ${op} failed.`, error?.recovery].filter(Boolean).join(" "));
  }
  return frame.response.result as T;
}

